### PR TITLE
Improve dpm-import and dpm-export

### LINKED ISSUE #605: Improve 2zhmc cpc dpm-import" handling for "preserve" flags
**Describe the bug**

Importing a DPM configuration happens as follows:

- existing configuration data is read from a json/yaml file by `zhmc`
- `zhmc` turns that configuration into a dictionary object
- `zhmc` sets the values for the `preserve-uris` and `preserve-wwpns` fields in said dict object
- the dict object is passed to the zhmcclient library call, which in turn uses the dict as request body for the actual WSAPI endpoint 

Past tooling allowed the user to include the to preservation fields directly in the configuration data file; or had the import tool add the fields before passing to the WSAPI endpoint.

The current implementation of `zhmc cpc dpm-import` doesn't allow for that; as the CLI code will always set the preserve fields (either with True or False); thus overwriting any existing field content originating in the configuration file itself.

**Expected behavior**

Either: document the current behavior explicitly in the CLI help 

Or: improve the CLI implementation to not necessarily overwrite flags already present in the configuration file.

**Environment information**

zhmc, version 1.10.1
zhmcclient, version 1.14.0

### PR DESCRIPTION
**Pull request by EdGue42**
Fixes #605 
- allow for uri/wwpn preserve and adapter mapping via config file
- include meta information in export data, dump/remove during import
- dump configuration after export / prior import
- move user confirmation accordingly

### CODE REVIEW

File: zhmccli/_cmd_cpc.py [hunk 1]
```diff
@@ -994,31 +1006,83 @@ def cmd_dpm_import(cmd_ctx, cpc_name, options):
...
-    if mapping_obj:
-        config_dict['adapter-mapping'] = \
-            convert_adapter_mapping(cmd_ctx, mapping_obj)
-    else:
-        if 'adapter-mapping' in config_dict:
-            del config_dict['adapter-mapping']
+    _log_mapping_and_overwrite_if_needed(cmd_ctx, config_dict, mapping_obj,
+                                         dpm_file, mapping_file)
+    _dump_config(config_dict)
+    cmd_ctx.spinner.stop()
```
Comment (EdGue42) [inline zhmccli/_cmd_cpc.py:1012]: not needed, stopped before!

File: zhmccli/_cmd_cpc.py [hunk 2]
```diff
@@ -20,20 +20,20 @@
 
 import io
 import json
+
+import click
 import yaml
 import yamlloader
-import click
+import zhmcclient
```
Comment (simon-spinner) [inline zhmccli/_cmd_cpc.py:27]: Why are all imports reordered?

File: zhmccli/_cmd_cpc.py [hunk 3]
```diff
@@ -349,28 +349,31 @@ def dpm_export(cmd_ctx, cpc, **options):
...
               'is read. Default: {d}.'.format(d=DEFAULT_DPM_FORMAT))
-@click.option('--preserve-uris', is_flag=True, required=False,
-              help='Controls whether the URIs and IDs of objects in the DPM '
-              'configuration file are preserved or ignored and newly '
-              'generated. Default: Ignored and newly generated.')
-@click.option('--preserve-wwpns', is_flag=True, required=False,
-              help='Controls whether the WWPNs of HBAs in the DPM '
-              'configuration file are preserved or ignored and newly '
-              'generated. Default: Ignored and newly generated.')
+@click.option('--preserve-uris/--generate-uris',
```
Comment (simon-spinner) [inline zhmccli/_cmd_cpc.py:360]: Don't you need to specify separate click options?

File: zhmccli/_cmd_cpc.py [hunk 4]
```diff
@@ -994,31 +1006,83 @@ def cmd_dpm_import(cmd_ctx, cpc_name, options):
...
+    cmd_ctx.spinner.stop()
+
+    if options["yes"]:
+        click.echo("already confirmed!")
+
+    message = "Are you sure you want to import the DPM configuration "
+    ("from {f} into CPC {c}, replacing its current DPM configuration?"
+     .format(c=cpc_name, f=dpm_file))
+
+    if not click.confirm(message):
```
Comment (simon-spinner) [inline zhmccli/_cmd_cpc.py:1021]: Does the --yes flag still work with this explicit confirmation?

File: zhmccli/_cmd_cpc.py (hunk 3 above)
Comment (EdGue42) [reply zhmccli/_cmd_cpc.py:369]: This is on purpose, so they end up "together" and one can check preserve-uris for "both" cases.

File: zhmccli/_cmd_cpc.py (hunk 4 above)
Comment (EdGue42) [reply zhmccli/_cmd_cpc.py]: with the updates I made, --yes

File: zhmccli/_cmd_cpc.py (hunk 2 above)
Comment (EdGue42) [reply zhmccli/_cmd_cpc.py:28]: fixed

File: docs/changes.rst [hunk 5]
```diff
@@ -65,6 +75,14 @@ Released: not yet
   This improved the time for listing partitions on a test system with many
   partitions from about 20 seconds to below 3 seconds.
 
+* cpc command dpm-export now prints a summary of the exported configuration
+  data.
+
+* cpc command dpm-imports now prints summary information regarding the
```
Comment (tobi1o24) [inline docs/changes.rst:81]:
```suggestion
* cpc command dpm-import now prints summary information regarding the
```

**Review by tobi1o24 — CHANGES_REQUESTED** [review]

File: docs/changes.rst [hunk 6]
```diff
@@ -29,6 +29,16 @@ Released: not yet
 
 **Incompatible changes:**
 
+* cpc command dpm-export now adds additional meta information to the export
+  data (can be suppressed using --exclude-extra-fields). Export configuration
+  files created with this version of zhmc can only be imported using this
+  version of zhmc, too!
```
Comment (simon-spinner) [inline docs/changes.rst:35]: ... and any newer version of zhmc.

File: zhmccli/_cmd_cpc.py [hunk 7]
```diff
@@ -323,6 +324,12 @@ def get_em_data(cmd_ctx, cpc):
               required=False, default=DEFAULT_DPM_FORMAT,
               help='Controls in which data format the DPM configuration file '
               'is written. Default: {d}.'.format(d=DEFAULT_DPM_FORMAT))
+@click.option('--exclude-extra-fields', '-e', is_flag=True, default=False,
```
Comment (simon-spinner) [inline zhmccli/_cmd_cpc.py:327]: Why not call it `--exclude-meta-information`? Would be good to clean up nomenclature with the description. "extra information", "additional information", "additional meta information" are all used.

Comment (EdGue42) [reply zhmccli/_cmd_cpc.py:327]: Excellent idea!

**Review by tobi1o24 — APPROVED** [review]

**Review by tobi1o24 — APPROVED** [review]

**Review by EdGue42 — COMMENTED** [review]
self review done ....

**Review by andy-maier — APPROVED** [review]

**Pull request merged**